File: app/src/main/java/io/github/sykoram/sis/MainActivity.java
package io.github.sykoram.sis;

import static io.github.sykoram.sis.MainActivityUtils.readFileToString;
import static io.github.sykoram.sis.MainActivityUtils.getPageIdentifier;
import static io.github.sykoram.sis.MainActivityUtils.isUrlAllowed;

import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.view.View;
import android.webkit.CookieManager;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.appcompat.app.AppCompatActivity;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import java.io.InputStream;

public class MainActivity extends AppCompatActivity {

    WebView webView;

    SwipeRefreshLayout refreshLayout;
    boolean pageFinished = true;

    /* The first path part after "is.cuni.cz/studium/" (ev. with ".php" removed):
         index, rozvrhng, predmety, term_st2, predm_st2, szz_st, zkous_st, anketa, dipl_st, grupicek, omne, role (...)
       This will be added as a class to <body>. */
    String pageIdentifier = "";

    WebViewClient webClient = new WebViewClient() {
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            if (isUrlAllowed(request.getUrl())) {
                return false;
            }
            Intent i = new Intent(Intent.ACTION_VIEW, request.getUrl());
            startActivity(i);
            return true;
        }

        @Override
        public void onPageStarted(WebView view, String url, Bitmap bitmap) {
            super.onPageStarted(view, url, bitmap);

            pageFinished = false;
            if (!refreshLayout.isRefreshing()) {
                refreshLayout.setRefreshing(true);
            }

            pageIdentifier = getPageIdentifier(url);
            try {
                view.evaluateJavascript("document.body.classList.add('" + pageIdentifier + "');", null);
                String styleScript = readFileToString(getResources().openRawResource(R.raw.script0));
                view.evaluateJavascript(styleScript, null);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        @Override
        public void onPageFinished(WebView view, String url) {
            super.onPageFinished(view, url);

            pageIdentifier = getPageIdentifier(url);
            view.evaluateJavascript("document.body.classList.add('" + pageIdentifier + "');", null);
            try (InputStream styleScriptIs = getResources().openRawResource(R.raw.script0);
                 InputStream scriptIs = getResources().openRawResource(R.raw.script1)) {
                view.evaluateJavascript(readFileToString(styleScriptIs), null);
                view.evaluateJavascript(readFileToString(scriptIs), null);
            } catch (Exception e) {
                e.printStackTrace();
            }

            pageFinished = true;
            onPageLoaded(view);
        }
    };

    WebChromeClient webChromeClient = new WebChromeClient() {
        @Override
        public void onProgressChanged(WebView view, int newProgress) {
            if (newProgress < 100 && !refreshLayout.isRefreshing()){
                refreshLayout.setRefreshing(true);
            }

            if (newProgress == 100 && pageFinished) {
                onPageLoaded(view);
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        webView = findViewById(R.id.webView);
        webView.setWebViewClient(webClient);
        webView.setWebChromeClient(webChromeClient);
        webView.getSettings().setCacheMode(WebSettings.LOAD_DEFAULT);
        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setAllowUniversalAccessFromFileURLs(true);
        webView.setVisibility(View.GONE);
        webView.loadUrl("https://is.cuni.cz/studium");

        refreshLayout = findViewById(R.id.swipeRefreshLayout);
        refreshLayout.setColorSchemeResources(R.color.primary_blue);
        refreshLayout.setEnabled(false);
    }

    public void onPageLoaded(WebView view) {
        refreshLayout.setRefreshing(false);
        if (view.getVisibility() == View.GONE) {
            view.setVisibility(View.VISIBLE);
        }
        CookieManager.getInstance().flush();
    }

    @Override
    public void onBackPressed() {
        if(webView.canGoBack()) {
            webView.goBack();
        } else {
            super.onBackPressed();
        }
    }

}

File: app/src/main/java/io/github/sykoram/sis/MainActivity.kt
package io.github.sykoram.sis

import android.app.DownloadManager
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.view.View
import android.webkit.*
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.net.toUri
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout

class MainActivity : AppCompatActivity() {
    lateinit var webView: WebView

    lateinit var refreshLayout: SwipeRefreshLayout

    var pageFinished = true

    val webClient: WebViewClient = object : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
            if (isUrlAllowed(request.url)) {
                return false
            }

            startActivity(Intent(Intent.ACTION_VIEW, request.url))
            return true
        }

        override fun onPageStarted(view: WebView, url: String, bitmap: Bitmap?) {
            super.onPageStarted(view, url, bitmap)

            pageFinished = false

            if (!refreshLayout.isRefreshing) {
                refreshLayout.isRefreshing = true
            }

            addPageIdentifiersToBody(view, url)

            resources.openRawResource(R.raw.script0)
                    .use { ins -> view.evaluateJavascript(readFileToString(ins), null) }
        }

        override fun onPageFinished(view: WebView, url: String) {
            super.onPageFinished(view, url)

            addPageIdentifiersToBody(view, url)

            resources.openRawResource(R.raw.script0)
                    .use{ ins -> view.evaluateJavascript(readFileToString(ins), null) }
            resources.openRawResource(R.raw.script1)
                    .use{ ins -> view.evaluateJavascript(readFileToString(ins), null) }

            pageFinished = true
            onPageLoaded(view)
        }

        override fun shouldInterceptRequest(view: WebView, request: WebResourceRequest): WebResourceResponse? {
            return replaceRequestWithAssets(request, assets) ?: super.shouldInterceptRequest(view, request)
        }
    }

    val webChromeClient: WebChromeClient = object : WebChromeClient() {
        override fun onProgressChanged(view: WebView, newProgress: Int) {
            if (newProgress < 100 && !refreshLayout.isRefreshing) {
                refreshLayout.isRefreshing = true
            }
            if (newProgress == 100 && pageFinished) {
                onPageLoaded(view)
            }
        }
    }

    val downloadListener: DownloadListener = object : DownloadListener {
        override fun onDownloadStart(url: String?, userAgent: String?, contentDisposition: String?, mimeType: String?, contentLength: Long) {
            val request = DownloadManager.Request(Uri.parse(url))
            request.addRequestHeader("cookie", CookieManager.getInstance().getCookie(url))
            request.addRequestHeader("User-Agent", userAgent)
            request.setMimeType(mimeType)
            request.setTitle(URLUtil.guessFileName(url, contentDisposition, mimeType))
            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
            request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, URLUtil.guessFileName(url, contentDisposition, mimeType))
            (getSystemService(DOWNLOAD_SERVICE) as DownloadManager).enqueue(request)

            Toast.makeText(applicationContext, R.string.downloading_file, Toast.LENGTH_LONG).show()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        // open drawer on first run
        val settings = getSharedPreferences("SharedPrefsFile", 0)
        if (settings.getBoolean("first_run", true)) {
            findViewById<DrawerLayout>(R.id.drawerLayout).openDrawer(GravityCompat.START, true)
            settings.edit().putBoolean("first_run", false).apply()
        }

        webView = findViewById(R.id.webView)
        webView.webViewClient = webClient
        webView.webChromeClient = webChromeClient
        webView.setDownloadListener(downloadListener)
        webView.settings.cacheMode = WebSettings.LOAD_DEFAULT
        webView.settings.javaScriptEnabled = true
        webView.visibility = View.GONE
        webView.loadUrl("https://is.cuni.cz/studium/index.php")

        refreshLayout = findViewById(R.id.swipeRefreshLayout)
        refreshLayout.setColorSchemeResources(R.color.primary_blue)
        refreshLayout.isEnabled = false
    }

    fun onPageLoaded(view: WebView) {
        refreshLayout.isRefreshing = false
        if (view.visibility == View.GONE) {
            view.visibility = View.VISIBLE
        }

        CookieManager.getInstance().flush()
    }

    override fun onBackPressed() {
        if (webView.canGoBack()) {
            webView.goBack()
        } else {
            super.onBackPressed()
        }
    }

    fun openPageExternally(view: View) {
        if (webView.url != null) {
            startActivity(Intent(Intent.ACTION_VIEW, webView.url!!.toUri()))
            findViewById<DrawerLayout>(R.id.drawerLayout).closeDrawers()
        }
    }

    fun refreshPage(view: View) {
        webView.reload()
        findViewById<DrawerLayout>(R.id.drawerLayout).closeDrawers()
    }
}